text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,749 @@
+
+In addition, it also stores references under their name, prefixed with `refs/`, as the prefix can never be part of a hex-formatted hash, thus automatically namespacing objects and refs.
+
+## Type safety
+
+An often brought up concern is the perceived type-safety loss by having a generic *object* as the central focus of this CEP. However, this is addressed by the following factors:
+
+* Every serialized object carries a type annotation in its hashed serialization, thus the kind of an object is always known at runtime.
+* Objects can be subjected to checked downcasts, e.g. an `as_block(&self) -> Option<&Block>` method that will only return a `Some` if the object in question is indeed a block.
+* The API for retrieving can conveniently allow specificying a type and roll up downcast errors into the retrieval error type. As an example, a method `get_object_from_storage<T>(&self) -> Result<T, Error>` can be called with `T = Object` for no downcasts or a `T = Block` if a block is expect, combining "not found" and "wrong errors" in `Error`.
+* Objects can still be passed around in their downcast form, only being upcast at IO barriers like storage or networking.
+
+Overall the amount of type safety retained or even added is firmly in the hand of the implementor of this CEP.
+
+## Global State
+
+The motivational section mentions that joining requires an up-to-date copy of the global state, which can only be built by replaying all deploys starting at Genesis.
+
+This, of course, is not entirely accurate - by attaching a hash of the global state's *root hash* (named *post state hash* when the hash is the result of the execution of a set of deploys from the current block), this process can be cut short:
+
+![Diagram depicting the relationship between previous post state hashes, blocks and deploys](../diagrams/0008-global-state.png)
+
+The proposed unified object model makes it trivial to download the global state by simply asking another node (or external caches, see [Future possibilities](#future-possibilities)) for the object referenced by `post_state_hash`. In the case depicted, the node which has Block, will download ParentBlock and the referenced GlobalState1, then apply both Deploys and check if the output is the same as the GlobalState2.
+
+## Loose ends
+
+Graph traversal is simple if a known starting point is available, e.g. when beginning from a trusted block that forms a root of the pseudo-tree used in an operation. Some objects are more "dynamic", they are added after the fact and point to existing objects, which must remain immutable.
+
+* Low levels of this "dynamicism" can be worked around by **introducing refs**. A newly created block has a parent dependency that points in the "wrong direction": From the perspective of a node that has a single trusted block as an ancestor of the new block, there is no connection. By keeping and updating refs like `refs/latest`, we keep multiple handles into the DAG, ensuring we do not miss these.
```

Comment:
Let's assume I am starting with a trusted block `T` and download the history, it would look like this

```
                 refs/latest
                   trusted
                      v
T_{-2} <-- T_{-1} <-- T
```

If I now mine a new block `T_{+1}`, the result is

```
                 refs/latest
                   trusted
                      v
T_{-2} <-- T_{-1} <-- T <-- T_{+1} 
```

The block is not "reachable" though, as I have not moved the trusted pointer. I'd suggest to introduce/move `refs/latest` here.

```
                
                  trusted   refs/latest
                      v        v
T_{-2} <-- T_{-1} <-- T <-- T_{+1} 
```